Add nested container type support for Arrays and Dictionaries

Sammyjroberts opened the pull request:
# Add nested container type support for Arrays and Dictionaries

**This is a draft PR for review and feedback before completion**

## Overview

This PR implements comprehensive support for nested container types in GDScript, enabling syntax like:
- `Array[Array[int]]`
- `Array[Dictionary[String, CustomClass]]` 
- `Dictionary[String, Array[Vector3]]`
- And deeper nesting up to 8 levels

## Technical Approach

### Core Changes

**Enhanced Type System** (`container_type_validate.h`):
- Extended `ContainerType` and `ContainerTypeValidate` with `nested_types` vectors
- Added recursive validation methods for nested arrays and dictionaries
- Implemented depth limiting (max 8 levels)
- Enhanced `can_reference()` and equality operators for nested type compatibility

**Array Implementation** (`array.cpp/h`):
- Added constructor overload accepting `Vector<ContainerTypeValidate>` for nested types
- Enhanced `assign()` and `append_array()` with nested type validation
- Implemented bidirectional conversion utilities between `ContainerType` ↔ `ContainerTypeValidate`
- Added `set_typed()` overload for nested type specification

**Parser Integration** (`gdscript_parser.cpp/h`):
- Modified type parsing to accept nested container syntax
- Added depth validation during parsing (8 level limit)
- Enhanced `DataType::to_string()` for proper nested type display
- Added helper method `get_container_type_depth()` for validation

**Type Analysis** (`gdscript_analyzer.cpp`):
- Updated array literal type inference to handle nested structures
- Enhanced type compatibility checking with recursive validation
- Improved element type updating for nested array literals

### Key Features
✅ **Parse-time Validation**: Syntax errors caught during compilation  
✅ **Recursive Operations**: Deep type checking for assignment, append, and other operations  
✅ **Backward Compatibility**: Existing single-level typed arrays continue to work unchanged  

## Example Usage

```gdscript
# Basic nested arrays
var nested_ints: Array[Array[int]] = [[1, 2], [3, 4]]
var matrix: Array[Array[Array[float]]] = [[[1.0, 2.0]], [[3.0, 4.0]]]

# Mixed nested containers  
var lookup: Dictionary[String, Array[Vector3]] = {
    "positions": [Vector3(1,2,3), Vector3(4,5,6)]
}
```

- *Bugsquad edit:* This PR closes https://github.com/godotengine/godot-proposals/issues/12224

Ivorforce commented:
Hi! You asked for feedback in RocketChat :) At a glance, I think this is the correct approach to go about adding this feature. As a note `TypedArray` and `TypedDictionary` will have to support the feature also, which may require you to work some C++ templates trickery. Plus, since core's current types system isn't first order yet, we may end up with some logic duplication because `Dictionary` uses two separate `ContainerTypeValidate` for key and value. Ideally, this PR should avoid as much duplication as possible. You'll have to figure that out as you work out the rest of the logic :) Good luck!

Maran23 commented:
This resolves): https://github.com/godotengine/godot-proposals/issues/12224
Happy to test as soon as this is somewhat stable!

Sammyjroberts commented:
@uwu-420 unfortunately I made it to the UI portion, and got stuck and busy with work.

The actual analyzer and parser are working (although the code needs love) but the UI (for example on @export of a nested container) was seriously broken.

Happy to support if possible once I have a bit more time in a week or so!

Ivorforce commented:
Well, the maintainer effort is primarily on the changes to the type system (afaik), but it's not out of the question that a well made PR could be considered for merge before that. It really depends on the PR.
This one cannot be considered until it is finished and out of draft.

Sammyjroberts commented:
Thanks for the update @kevinlam508 I am working on this again, and expect to push a diff soon!

Sammyjroberts committed:
GDScript: Support nested typed collections

Allows declaring and using arbitrarily nested typed containers such as
`Array[Array[int]]`, `Dictionary[String, Array[int]]`, and
`Dictionary[String, Dictionary[String, int]]` up to a fixed depth.

Parser/analyzer:
- `parse_type` accepts nested type arguments and rejects types that
  exceed MAX_CONTAINER_NESTING_DEPTH (8).
- `update_array_literal_element_type` and
  `update_dictionary_literal_element_type` recurse into nested
  container element types so inner literals are typed correctly.
- `check_type_compatibility` recurses into container element types for
  both ARRAY and DICTIONARY so nested typed assignments type-check.
- `@export` on a nested container builds a full recursive
  PROPERTY_HINT_TYPE_STRING via `build_container_type_hint`, which
  emits the type id of each level joined by `/` for hint and `:` for
  the element descriptor, and `;` for dictionary key/value.

Runtime (Array / Dictionary / ContainerTypeValidate):
- `ContainerType` and `ContainerTypeValidate` carry a `nested_types`
  vector describing the type structure beneath the outer container.
- `Array::set_typed` and `Dictionary::set_typed` gain overloads that
  accept the nested type vector; the runtime operators (`assign`,
  `append_array`, `merge`, `keys`, `values`) propagate and validate
  the nested type information.
- `validate_nested_array` / `validate_nested_dictionary` walk the
  element(s) recursively, coercing when possible (e.g. int -> float).
  The dictionary validator iterates a const source via KeyValue and
  builds a fresh Dictionary; it only writes back when something was
  actually coerced, preserving COW and leaving the source untouched
  on error. Duplicate-key detection after coercion is enforced.
- `MAX_CONTAINER_NESTING_DEPTH = 8` lives in
  `container_type_validate.h` and is shared by the parser, the
  runtime set_typed paths, and the validator.

Inspector UI:
- Port of #117529: `EditorPropertyDictionary::setup` splits the hint
  string on the first `;` only so nested dictionary hint strings are
  passed down intact.
- `EditorPropertyArray::setup` and `EditorPropertyDictionary::setup`
  accept numeric type ids in the no-colon branch (including `0` as
  the NIL / Variant wildcard), in addition to type names.
- `update_property` builds a recursive display label via a new
  `_get_nested_container_display_name` helper so nested exports show
  as e.g. `Array[Array[int]]` or `Dictionary[String, Array[int]]`.

Tests:
- `analyzer/features/nested_typed_array.gd`
- `analyzer/features/nested_typed_dictionary.gd`
- `analyzer/errors/nested_typed_array_incompatible_literal.gd`
- `parser/errors/nested_typed_collection_depth_exceeded.gd`
- `runtime/features/nested_typed_collections.gd`

Known follow-up (not covered here): resizing a nested typed array in
the inspector leaves new element slots as Nil until the user clicks
into them. This belongs in `Array::resize`'s default-element init
path and is tracked for a separate change.

pl4ttenbau commented:
Certainly out of draft

will be merged in 2032 i'm sure.
